lib/carrier/messaging/tracker.ex

```diff
@@ -0,0 +1,179 @@
+            {maybe_unmonitor_subscriber(%{tracker | subscriptions: Map.put(subs, topic, {matcher, updated})}, subscriber), true}
+        end
+      nil ->
+        {tracker, false}
+    end
+  end
+
+  @spec find_reply_subscriber(Tracker.t, String.t) :: pid() | nil
+  def find_reply_subscriber(%__MODULE__{reply_endpoints: reps}, topic) do
+    Map.get(reps, topic)
+  end
+
+  @spec find_subscribers(Tracker.t, String.t) :: [] | [pid()]
+  def find_subscribers(%__MODULE__{subscriptions: subs}, topic) do
+    Enum.reduce(subs, [], &(find_matching_subscriptions(&1, topic, &2)))
+  end
+
+  @spec del_subscriber(Tracker.t, pid()) :: Tracker.t
+  def del_subscriber(tracker, subscriber) do
+    tracker
+    |> del_reply_endpoint(subscriber)
+    |> del_all_subscriptions(subscriber)
+    |> unmonitor_subscriber(subscriber)
+  end
+
+  @spec unused?(Tracker.t) :: boolean()
+  def unused?(%__MODULE__{reply_endpoints: reps, subscriptions: subs, monitors: monitors}) do
+    Enum.empty?(reps) and Enum.empty?(subs) and Enum.empty?(monitors)
+  end
+
+  @spec get_and_reset_unused_topics(Tracker.t) :: {Tracker.t, [String.t]}
+  def get_and_reset_unused_topics(tracker) do
+    {%{tracker | unused_topics: []}, tracker.unused_topics}
+  end
+
+  defp make_reply_endpoint_topic() do
+    id = UUID.uuid4(:hex)
+    "carrier/call/reply/#{id}"
+  end
+
+  defp find_matching_subscriptions({_, {matcher, subscribed}}, topic, accum) do
+    if Regex.match?(matcher, topic) do
+      accum ++ subscribed
+    else
+      accum
+    end
+  end
+
+  defp subscription_matcher(sub_topic) do
+    regex = case String.slice(sub_topic, -2, 2) do
+              "/+" ->
+                "^#{String.slice(sub_topic, 0, String.length(sub_topic) - 1)}[a-zA-Z0-9_]+$"
+              "/*" ->
+                "^#{String.slice(sub_topic, 0, String.length(sub_topic) - 1)}.*"
+              _ ->
+                "^#{sub_topic}$"
+            end
+    Regex.compile!(regex)
+  end
+
+  def maybe_monitor_subscriber(%__MODULE__{monitors: monitors}=tracker, subscriber) do
+    case Map.get(monitors, subscriber) do
+      nil ->
+        mref = :erlang.monitor(:process, subscriber)
+        %{tracker | monitors: Map.put(monitors, subscriber, mref)}
+      _ ->
+        tracker
+    end
+  end
+
+  def maybe_unmonitor_subscriber(%__MODULE__{monitors: monitors}=tracker, subscriber) do
+    if has_subscriptions?(tracker, subscriber) do
+      tracker
+    else
+      if Map.has_key?(monitors, subscriber) do
+        unmonitor_subscriber(tracker, subscriber)
+      else
+        tracker
+      end
+    end
+  end
+
+  def unmonitor_subscriber(%__MODULE__{monitors: monitors}=tracker, subscriber) do
+    mref = Map.fetch!(monitors, subscriber)
+    :erlang.demonitor(mref, [:flush])
+    %{tracker | monitors: Map.delete(monitors, subscriber)}
+  end
+
+  defp has_subscriptions?(tracker, subscriber) do
+    if Enum.reduce_while(tracker.reply_endpoints, false, &(has_endpoint(&1, subscriber, &2))) do
```

Comment:
Since `reply_endpoints` is bidirectional, could this be replaced with `Map.has_key(reply_endpoints, subscriber)`?